### DISCUSSION TITLE
What to do Docker Hub registry returns 429 Too Many Requests errors?

### DISCUSSION THREAD

**Discussion by gerhard** [dagger/dagger#2997] [Help]
I was running `dagger do test` a few times and have hit the `429 Too Many Requests` error:

<img width="1737" alt="image" src="https://user-images.githubusercontent.com/3342/188471397-d70cc848-8ea1-413d-893e-2d3e0aed2912.png">

I have confirmed that Docker Desktop is logged in via:

```sh
curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest
```

* FYI https://docs.docker.com/docker-hub/download-rate-limit/#how-can-i-check-my-current-rate
* FYI2 https://docs.gitlab.com/ee/user/packages/container_registry/#authenticate-by-using-gitlab-cicd

Docker login is obviously not helpful, since the images are pulled by BuildKit, not Docker.

So what should we do when BuildKit hits the `429 Too Many Requests` error? cc @sipsma @aluzzardi 

---

I am attaching the asciinema file in case you want to replay what I did with `asciinema play dagger-do-test-429-too-many-requests.cast` 👉 [dagger-do-test-429-too-many-requests.cast.zip](https://github.com/dagger/dagger/files/9490987/dagger-do-test-429-too-many-requests.cast.zip)

**Comment by marcosnils** [comment]
Hmm good catch. I don't think there's much we can do here but to pass the registry credentials to the buildkit builder. A workaround that I can come up with in the meantime is creating the builder manually through `docker run` and pass a buildkit config with the corresponding registry access. 

I wouldn't make any changes to the dagger code since it's very likely that we'll change it soon when working on cloak's embedded engine.

**Comment by vito** [comment]
(:wave:)

> Docker login is obviously not helpful, since the images are pulled by BuildKit, not Docker.

It's possible to load the Docker auth config from `~/.docker/config.json` (client-side) and pass it to Buildkit via `buildkit/client.SolveOpt.Session` if we want this to magically work.

Some pseudocode:

```go
import (
	dockerconfig "github.com/docker/cli/cli/config"
	kitdclient "github.com/moby/buildkit/client"
	"github.com/moby/buildkit/session/auth/authprovider"
)

func main() {
	_, err := runtime.Client.Solve(ctx, def, kitdclient.SolveOpt{
		Session: []session.Attachable{
			authprovider.NewDockerAuthProvider(dockerconfig.LoadDefaultConfigFile(os.Stderr)),
		},
	}, nil)
}
```

Whether that's advisable is another story; it might be wise to configure registry auth directly in Dagger somehow rather than forcing a direct coupling to Docker. But it's nice that we can pass it along on a per-request basis rather than configuring it Buildkit-side. :)

**Reply by marcosnils** [reply]
> (:wave:)

Hey vito! good to see you here!


> But it's nice that we can pass it along on a per-request basis rather than configuring it Buildkit-side. :)

definitely! it's nice to have a fine-grained way to pass this into the specific session. I guess coming up with the proper API / UX to configure it will be the "difficult" thing to do now. Do I sense a `dagger.json` conf coming?

**Reply by vito** [reply]
> Hey vito! good to see you here!

Thanks, happy to be here! :smile: 

> I guess coming up with the proper API / UX to configure it will be the "difficult" thing to do now. Do I sense a dagger.json conf coming?

Yup, that's what I'm thinking.

Another way to look at this could be integrating it with secret management. It looks like the docs [already cover](https://docs.dagger.io/1244/docker/#dockerpull) how to configure credentials for pulling images, which I presume uses something like the code above under the hood, but I would expect to only need to do that for private images repositories/registries. Having to set creds everywhere just to get the 100->200 req/6h rate limit bump would be pretty messy. :thinking: 

So yeah, that brings me back to some sort of per-registry default auth on the `dagger` user's local machine, since I wouldn't want to have to modify project code for this, and different users of the same project may have different preferences for how they manage their creds (e.g. 1password,  brain, post-its).

**Comment by grouville** [comment]
Hi, here is a PR that could be a temporary solution to that: https://github.com/dagger/dagger/pull/3002

**Comment by 3nprob** [comment]
I opened an issue with some suggestions for improving the situation:
- https://github.com/dagger/dagger/issues/11125
